### PR TITLE
feat(bot): add discography mode to /artist command

### PR DESCRIPTION
**Pull request by LucasSantana-Dev**
## Why

Operator: "the artist command has a limit param that's not respected, if I set an artist most of the times I want to hear their entire discography from most famous to least and just stop when I say so, add a parameter for that."

Root cause of the limit complaint: `discord-player-spotify`'s `SpotifyAPI.search()` hardcodes `limit=10` server-side (both the credentialed and anonymous GraphQL code paths — verified in `node_modules/discord-player-spotify/dist/index.js`). `/artist`'s plain flow draws from that capped pool before the artist-match filter even runs, so `limit:15`/`limit:20` could never actually return more than ~10 candidates. Tracked separately in #2304 since it's a third-party dependency limitation, not something this PR patches.

## What

Adds a `discography` boolean option to `/artist`. When set:
- Resolves the artist via Spotify's artist search (exact-name match preferred).
- Pulls the artist's real top-tracks (`/v1/artists/{id}/top-tracks`, Spotify's own popularity ranking — genuinely most-famous-first) and up to 25 albums/singles (`/v1/artists/{id}/albums`, newest first, deduped by name to skip deluxe/reissue duplicates).
- Resolves each through discord-player exactly like `/album` already does (same URL-based search pattern), dedupes the combined list by title+artist so a top track doesn't get requeued from its parent album, and caps at 300 total tracks as a safety ceiling against extremely prolific catalogs.
- Ignores the `limit` option when `discography` is true.
- Stop via the existing `/skip` / `/stop` commands.

Reuses existing shared Spotify Web API plumbing (`getSpotifyClientToken`, `searchSpotifyArtists`, the `SpotifyAlbumMatch` type/mapper from `/album`'s resolver) rather than introducing a new auth or HTTP pattern.

## Test plan

- [x] `npx tsc --noEmit` clean across `packages/shared` and `packages/bot`
- [x] New unit tests: `packages/shared/src/utils/spotify/artistApi.spec.ts` (top-tracks mapping, album pagination/dedupe/sort/cap, graceful failure) and `packages/bot/src/functions/music/commands/artistDiscography.spec.ts` (no-credentials, no-match, exact-match preference, dedupe across top-tracks/albums, partial-album-failure resilience)
- [x] Existing `/artist` and `/album` suites still pass unmodified (`artist.spec.ts` fixture updated with a `getBoolean` mock)
- [x] `eslint` clean (0 errors) on all touched files

---
## Summary by cubic
Adds a `discography` option to `/artist` so users can queue an artist's whole catalog, most famous first, instead of a fixed track count. `limit` is ignored when `discography` is set, and playback stops with the existing `/skip` or `/stop` commands.

**Details**
- Resolves the artist via Spotify search, preferring exact-name matches.
- Pulls the artist's real top tracks and up to 25 albums/singles, deduped by name and by title+artist, and caps the queue at 300 tracks.
- Resolves tracks through discord-player the same way `/album` does and reuses existing Spotify Web API helpers.
- The `limit` option's ~10 track ceiling is a third-party `discord-player-spotify` limitation tracked in #2304, not changed here.
- The CI smoke test now fails the build if built bot modules crash at load time, a class of bug unit tests and `tsc` can't catch.

<sup>Written for commit 04c30aa2e7dcb986c255c849bc72125a3b181d48. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/LucasSantana-Dev/Lucky/pull/2305?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: packages/bot/src/functions/music/commands/artistDiscography.ts [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    client: CustomClient,
+    interaction: ChatInputCommandInteraction,
+    url: string,
+): Promise<ResolvedTrack[]> {
+    try {
+        const result = await client.player.search(url, {
+            requestedBy: interaction.user,
+            searchEngine: QueryType.AUTO,
+        })
+        return result?.tracks ?? []
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/functions/music/commands/artistDiscography.ts:63]: Not a bug — verified against the actual library source (node_modules/discord-player-spotify/dist/index.js:455-483). Album-URL resolution fully paginates (limit:50/page, loops `while (allItems.length < totalCount)`) and returns the complete tracklist regardless of size. The 10-item cap cubic describes only applies to discord-player-spotify's generic search-query path (SpotifyAPI.search()), not direct album URLs — this is exactly why resolveAlbumTracks passes an album URL rather than a search query (see also #2304, filed earlier this session for the plain /artist limit bug caused by that same generic-search cap).

File: packages/bot/src/functions/music/commands/artistDiscography.spec.ts [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+describe('handleArtistDiscography', () => {
+    let addTrack: jest.Mock
+    let play: jest.Mock
+    let search: jest.Mock
+    let interaction: { guildId: string; user: { id: string }; channel: object }
+    let client: { player: { search: jest.Mock; play: jest.Mock } }
+
+    beforeEach(() => {
+        jest.clearAllMocks()
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/functions/music/commands/artistDiscography.spec.ts:83]: Fixed — resetting the four module-level Spotify mocks in beforeEach via mockReset(), not a blanket jest.resetAllMocks() (verified that breaks 5/6 tests here since createErrorEmbed/createSuccessEmbed's factory-time implementations only run once per file and never get restored). See 29ac352a.

File: packages/shared/src/utils/spotify/artistApi.ts [hunk 3]
```diff
@@ -166,3 +167,125 @@ export async function getSpotifyRelatedArtists(
...
+        artists?: { name?: string }[]
+        external_urls?: { spotify?: string }
+    }
+    const albums: RawAlbum[] = []
+    let url: string | null =
+        `https://api.spotify.com/v1/artists/${encodeURIComponent(artistId)}/albums?include_groups=album,single&market=US&limit=50`
+    try {
+        // Spotify caps each page at 50; stop once we have enough raw candidates
+        // to dedupe/sort/slice down to maxAlbums, or the artist runs out of pages.
+        while (url && albums.length < maxAlbums * 3) {
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/shared/src/utils/spotify/artistApi.ts]: Fixed — dedup now runs inside the pagination loop, keyed on distinct-album count instead of raw-item count, so duplicate-heavy pages no longer cut the fetch short. See 29ac352a.

File: packages/bot/src/functions/music/commands/artistDiscography.ts [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+            })
+            return
+        }
+
+        for (const track of tracks.slice(1)) {
+            track.requestedBy = interaction.user
+            queue.addTrack(track)
+        }
+
+        moveUserTrackToPriority(queue, playResult.track)
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/functions/music/commands/artistDiscography.ts]: Fixed — added the null guard exactly as suggested. Also had to fix the one test this exposed: it asserted moveUserTrackToPriority was called while play resolved a null track by default, so the assertion was passing on a call with no track — updated it to resolve a real track. See 29ac352a.

File: packages/bot/src/functions/music/commands/artistDiscography.ts [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    } catch (error) {
+        warnLog({
+            message: 'Discography album resolve failed',
+            data: { url, error: String(error) },
+        })
+        return []
+    }
+}
+
+export async function handleArtistDiscography({
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/functions/music/commands/artistDiscography.ts:73]: Acknowledged — the fan-out is the orchestration logic (token fetch, artist match, top-tracks + albums, resolve, dedupe, play/queue) that this command necessarily coordinates; splitting it wouldn't reduce actual complexity, just relocate it behind more indirection. Leaving as-is.

[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/functions/music/commands/artistDiscography.ts:73]: Acknowledged — the fan-out is the orchestration logic (token fetch, artist match, top-tracks + albums, resolve, dedupe, play/queue) that this command necessarily coordinates; splitting it wouldn't reduce actual complexity, just relocate it behind more indirection. Leaving as-is.

**Comment by LucasSantana-Dev** [comment]
graphify's coupling-delta findings on `handleArtistDiscography()` are expected for a command handler orchestrating several collaborators (Spotify lookups, queueing, embeds). No correctness issue — resolving as intentional architecture.
